### ISSUE TITLE
"Step must only increase in log calls" when adding W&B logger after some training

### ISSUE THREAD

**Issue by jonashaag** [wandb/client#1626]
**Describe the bug**

First time W&B user here. I added the `WandbLogger` to my PyTorch-Lightning project using the default settings, ie. `Trainer(logger=WandbLogger())`. Note that I'm resuming training from a checkpoint (didn't use W&B logger before).

I'm getting lots of warnings like `wandb: WARNING Step must only increase in log calls.  Step 47 < 270952; dropping {'loss': -2.309272527694702}` and no metrics are reported. Other data like GPU utilisation is collected correctly.

**Comment by borisdayma** [comment]
How do you log your metrics? Do you use `self.log`?
Also do you have some validation metrics logged? If so are you trying to log them at the batch level or are they aggregated and logged at the end?

**Comment by ternaus** [comment]
Same story.

Pytorch Lightning: `1.1.2rc1`
Wandb: `0.10.12`

No validation.

Logging with `self.log(...)`

Get errors like: `wandb: WARNING Step must only increase in log calls.  Step 49 < 60049; dropping {'l1_step': 0.10998408496379852, 'feature_matching_step': 0.19095762073993683, 'lpips_step': 0.2102833241224289, 'gan_step': 0.7236597537994385, 'total_loss_step': 4.846053123474121, 'discriminator_step': 0.1621440052986145, 'discriminator_true_step': 0.2694278657436371, 'discriminator_fake_step': 0.054860156029462814}`

**Comment by borisdayma** [comment]
It is most likely due to the logging of validation metrics at different steps.
Can you try to comment them out?

If it works, you can use directly `wandb.log(my_dict, commit=False)`

I'm looking into a way to allow more flexibility with logging at different stages.

**Comment by ternaus** [comment]
I do not have a validation step.

I think the issue is on the Pytorch Lightning side. I do not have this issue with version 1.0.8

**Comment by borisdayma** [comment]
You could look at your logged graphs and see what metrics are logged with such a high step.

**Comment by jonashaag** [comment]
> How do you log your metrics? Do you use `self.log`?
> Also do you have some validation metrics logged? If so are you trying to log them at the batch level or are they aggregated and logged at the end?

I use [Asteroid](https://github.com/asteroid-team/asteroid/) with default settings. Here's the log code:

https://github.com/asteroid-team/asteroid/blob/1dc1c130f5c8a108fa1e285ed8c194e11ad1ed27/asteroid/engine/system.py#L116
https://github.com/asteroid-team/asteroid/blob/1dc1c130f5c8a108fa1e285ed8c194e11ad1ed27/asteroid/engine/system.py#L128
https://github.com/asteroid-team/asteroid/blob/1dc1c130f5c8a108fa1e285ed8c194e11ad1ed27/asteroid/engine/system.py#L130-L134

I use `Trainer.resume_from_checkpoint`, and no additional metrics. Train loss is logged every batch, and validation loss logged every validation epoch.

**Comment by Tiiiger** [comment]
encountered the same issue. @jonashaag did you figure out how to deal with this?

**Comment by borisdayma** [comment]
I think the issue is at [this line](https://github.com/asteroid-team/asteroid/blob/1dc1c130f5c8a108fa1e285ed8c194e11ad1ed27/asteroid/engine/system.py#L134) and that maybe there's an issue with the step given.

Can you replace this with a call to `self.log("hp_metric", hp_metric)`?

If this does not work, you can also use https://github.com/PyTorchLightning/pytorch-lightning/pull/5351 with `sync_step=False`. The advantage is that you will also be able to see from your wandb graphs at what moment the step increases.
You will also be able to set up your own x-axis on your graphs if you want.

**Comment by jonashaag** [comment]
> encountered the same issue. @jonashaag did you figure out how to deal with this?

I’m dealing with it by not using W&B anymore ^^

**Comment by borisdayma** [comment]
This was fixed with #5931.
You can now just log at any time with `self.log('my_metric', my_value)` and you won't have any dropped value.
Just choose your x-axis appropriately in the UI (whether `global_step` or just the auto-incremented `step`).

**Comment by yllgl** [comment]
> This was fixed with [PyTorchLightning/pytorch-lightning#5931](https://github.com/PyTorchLightning/pytorch-lightning/pull/5931).
> You can now just log at any time with `self.log('my_metric', my_value)` and you won't have any dropped value.
> Just choose your x-axis appropriately in the UI (whether `global_step` or just the auto-incremented `step`).

@vanpelt @borisdayma how to log at the different stages? here is my code:
```python
import os
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning.loggers import WandbLogger
from pytorch_lightning import LightningModule, Trainer
import numpy as np
import wandb
class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)
        self.val_step=0
    def forward(self, x):
        return self.layer(x)
    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss,on_step = True, on_epoch = True, prog_bar = True, logger = True)
        return {"loss": loss}
    def validation_step(self,batch,batch_idx):
        test_img = np.random.randint(256,size=(512,512))
        self.logger.experiment.log({
            f"val_epoch{self.current_epoch}": wandb.Image(test_img, caption='caption')
            },step=self.val_step)
        self.val_step+=1
    def validation_epoch_end(self,ouputs):
        self.val_step=0
    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


train_data = DataLoader(RandomDataset(32, 64), batch_size=8)  # each gpu gets a batch size 4
val_data = DataLoader(RandomDataset(32, 64), batch_size=8)
logger = WandbLogger(project='test')
model = BoringModel()
trainer = Trainer(
    max_epochs=2,logger=logger,overfit_batches=2,weights_summary=None
)
trainer.fit(model, train_dataloader=train_data,val_dataloaders=val_data)
```

But it just output:
```python
wandb: WARNING Step must only increase in log calls.  Step 0 < 2; dropping {'val_epoch0': <wandb.sdk.data_types.Image object at 0x2ac216b97d00>}.
wandb: WARNING Step must only increase in log calls.  Step 1 < 2; dropping {'val_epoch0': <wandb.sdk.data_types.Image object at 0x2ac216b97e80>}.
wandb: WARNING Step must only increase in log calls.  Step 0 < 3; dropping {'val_epoch1': <wandb.sdk.data_types.Image object at 0x2ac21c785e50>}.
wandb: WARNING Step must only increase in log calls.  Step 1 < 3; dropping {'val_epoch1': <wandb.sdk.data_types.Image object at 0x2ac21c785700>}.
```
How could I log at the different stages?

**Comment by vanpelt** [comment]
Hey @jonashaag instead of passing the "step" argument, you can log separate x-axis as separate metrics.  I.E.

```
self.logger.experiment.log({
            "val_epoch": wandb.Image(test_img, caption='caption'),
            "val_step": self.val_step}, step=self.current_epoch)
```

Then in the UI you can choose a custom x-axis if you need to plot metrics against "val_step" for instance.  You also shouldn't be adding the "epoch" number to the key as this will create an unbounded number of charts.  You can remove `step=self.current_epoch` in the above but it's an optimization to ensure all metrics for a given epoch are logged together.
